crates/matrix-sdk-ui/src/sync_service.rs

```diff
@@ -105,71 +178,106 @@ impl SyncTaskSupervisor {
+                        warn!(?report, "unable to stop room list service: {err:#}");
+                    }
+
+                    if report.has_expired {
+                        room_list_service.expire_sync_session().await;
+                    }
                 }
 
-                if report.has_expired {
-                    room_list_service.expire_sync_session().await;
+                if let Err(err) = room_list_task.await {
+                    error!("when awaiting room list service: {err:#}");
                 }
-            }
 
-            if let Err(err) = room_list_task.await {
-                error!("when awaiting room list service: {err:#}");
-            }
+                if stop_encryption {
+                    if let Err(err) = encryption_sync.stop_sync() {
+                        warn!(?report, "unable to stop encryption sync: {err:#}");
+                    }
 
-            if stop_encryption {
-                if let Err(err) = encryption_sync.stop_sync() {
-                    warn!(?report, "unable to stop encryption sync: {err:#}");
+                    if report.has_expired {
+                        encryption_sync.expire_sync_session().await;
+                    }
                 }
 
-                if report.has_expired {
-                    encryption_sync.expire_sync_session().await;
+                if let Err(err) = encryption_sync_task.await {
+                    error!("when awaiting encryption sync: {err:#}");
                 }
-            }
 
-            if let Err(err) = encryption_sync_task.await {
-                error!("when awaiting encryption sync: {err:#}");
-            }
-
-            if report.is_error {
-                state.set(State::Error);
-            } else if matches!(report.origin, TerminationOrigin::Supervisor) {
-                state.set(State::Idle);
-            } else {
-                state.set(State::Terminated);
+                if report.is_error {
+                    if offline_mode {
+                        state.set(State::Offline);
+
+                        let client = room_list_service.client();
+
+                        if let Some(report) = Self::offline_check(client, &mut receiver).await {
+                            if report.is_error {
+                                state.set(State::Error);
+                                break;
+                            } else {
+                                state.set(State::Idle);
+                                break;
+                            }
+                        } else {
+                            state.set(State::Running);
+                        }
+                    } else {
+                        state.set(State::Error);
+                        break;
+                    }
+                } else if matches!(report.origin, TerminationOrigin::Supervisor) {
+                    state.set(State::Idle);
+                    break;
+                } else {
```

Comment:
nit: else after break. We can get rid of the `else` here.

Reply:
This one doesn't work, because not every branch has a break or continue, which means that this will be executed every time if we remove the else making the loop always execute only once.